### ISSUE TITLE
slow http response

### ISSUE THREAD

**Issue by audricbizouarn** [ClickHouse/ClickHouse#17403]
Hello,
we have a cluster of 4 servers (2 shards, 2 replicas). version 20.9.2.20
Sometimes the http response time on one of the servers becomes very long.
`
client: time curl http: // srvA: 8123
Ok.
real 0m12.556s
user 0m0.007s
sys 0m0.000s
`

`
client: time curl http: // srvB: 8123
Ok.
real 0m0.055s
user 0m0.006s
sys 0m0.000s
`

If we restart the service on srvA, the problem disappears.

Where can the problem come from? What can we check on the server when this happens?
Thank you.

**Comment by den-crane** [comment]
It could be anything. Network issue. Slow query execution. System swapping.
Why do you think that the issue is in http ?
How we can reproduce this issue?

**Comment by audricbizouarn** [comment]
if I ping the server at the same time, there is no problem.
And if I query clickhouse in tcp mode I don't have a problem either.
I don't see swap with htop.
I sometimes have no process running (SHOW PROCESSLIST) and yet I have the problem

**Comment by audricbizouarn** [comment]
It happens randomly. We don't understand how to reproduce it.

**Comment by den-crane** [comment]
Please show clickhouse-server logs for this query.

Try to ping **-s 1500**

**Comment by audricbizouarn** [comment]
```
time curl 'http://user:password@srvA:8123/?query=SELECT%201'
1

real	0m5.775s
user	0m0.013s
sys	0m0.000s
```

```
2020.11.30 08:35:18.565320 [ 24498 ] {} <Trace> HTTPHandler-factory: HTTP Request for HTTPHandler-factory. Method: GET, Address: [::ffff:ip]:49338, User-Agent: curl/7.35.0, Content Type: , Transfer Encoding: identity
2020.11.30 08:35:18.565435 [ 24498 ] {} <Trace> DynamicQueryHandler: Request URI: /?query=SELECT%201
2020.11.30 08:35:18.565579 [ 24498 ] {2b5748b8-eed1-4f1d-a9fe-53497ef3adfa} <Debug> executeQuery: (from [::ffff:ip]:49338, user: radio) SELECT 1 
2020.11.30 08:35:18.565735 [ 24498 ] {2b5748b8-eed1-4f1d-a9fe-53497ef3adfa} <Trace> ContextAccess (radio): Access granted: SELECT(dummy) ON system.one
2020.11.30 08:35:18.565767 [ 24498 ] {2b5748b8-eed1-4f1d-a9fe-53497ef3adfa} <Trace> InterpreterSelectQuery: FetchColumns -> Complete
2020.11.30 08:35:18.565984 [ 24498 ] {2b5748b8-eed1-4f1d-a9fe-53497ef3adfa} <Information> executeQuery: Read 1 rows, 1.00 B in 0.000372124 sec., 2687 rows/sec., 2.62 KiB/sec.
2020.11.30 08:35:18.566024 [ 24498 ] {2b5748b8-eed1-4f1d-a9fe-53497ef3adfa} <Information> DynamicQueryHandler: Done processing query
2020.11.30 08:35:18.566035 [ 24498 ] {2b5748b8-eed1-4f1d-a9fe-53497ef3adfa} <Debug> MemoryTracker: Peak memory usage (for query): 0.00 B.

```


```
ping -s 1500 srvA
PING srvA (srvA) 1500(1528) bytes of data.
1508 bytes from srvA: icmp_seq=1 ttl=64 time=0.674 ms
1508 bytes from srvA: icmp_seq=2 ttl=64 time=0.336 ms
1508 bytes from srvA: icmp_seq=3 ttl=64 time=0.265 ms
1508 bytes from srvA: icmp_seq=4 ttl=64 time=7.33 ms
1508 bytes from srvA: icmp_seq=5 ttl=64 time=0.335 ms
1508 bytes from srvA: icmp_seq=6 ttl=64 time=2.66 ms
1508 bytes from srvA: icmp_seq=7 ttl=64 time=0.254 ms
1508 bytes from srvA: icmp_seq=8 ttl=64 time=0.592 ms
1508 bytes from srvA: icmp_seq=9 ttl=64 time=1.87 ms
1508 bytes from srvA: icmp_seq=10 ttl=64 time=0.248 ms
1508 bytes from srvA: icmp_seq=11 ttl=64 time=0.249 ms
1508 bytes from srvA: icmp_seq=12 ttl=64 time=0.933 ms
1508 bytes from srvA: icmp_seq=13 ttl=64 time=2.24 ms
1508 bytes from srvA: icmp_seq=14 ttl=64 time=0.321 ms
1508 bytes from srvA: icmp_seq=15 ttl=64 time=0.298 ms
1508 bytes from srvA: icmp_seq=16 ttl=64 time=0.460 ms
1508 bytes from srvA: icmp_seq=17 ttl=64 time=0.249 ms
1508 bytes from srvA: icmp_seq=18 ttl=64 time=0.256 ms
1508 bytes from srvA: icmp_seq=19 ttl=64 time=0.265 ms
```

**Comment by den-crane** [comment]
Can you show an output of `curl -vvv 'http://user:password@srvA:8123/?query=SELECT%201'

**Comment by gj-zhang** [comment]
> Hello,
> we have a cluster of 4 servers (2 shards, 2 replicas). version 20.9.2.20
> Sometimes the http response time on one of the servers becomes very long.
> `client: time curl http: // srvA: 8123 Ok. real 0m12.556s user 0m0.007s sys 0m0.000s`
> 
> `client: time curl http: // srvB: 8123 Ok. real 0m0.055s user 0m0.006s sys 0m0.000s`
> 
> If we restart the service on srvA, the problem disappears.
> 
> Where can the problem come from? What can we check on the server when this happens?
> Thank you.

hi i have the same problme.  do you resolve?

**Comment by Algunenano** [comment]
Just a note in case somebody finds this ticket again. The issue was due to a race condition in the TCP Dispatcher were it would "leak" in its thread counter. At some point the server would think it could not request any more threads from the pool and new connections would be queued up, leading to the extra latency seem here.

This was fixed in https://github.com/ClickHouse-Extras/poco/pull/28 so it's fixed in 21.01 and it was backported to 21.12 (https://github.com/ClickHouse/ClickHouse/pull/17542). A workaround (with other consequences) in older versions would be to increase `max_connections` but that only increases how long it takes the problem to appear.
